overlord/snapstate/backend/link.go

```diff
@@ -63,19 +63,90 @@ type LinkContext struct {
 	SkipBinaries bool
 }
 
-func updateCurrentSymlinks(info *snap.Info) (e error) {
+func cleanPreviousSymlinks(info *snap.Info) {
+	if info.Type() != snap.TypeSnapd {
+		return
+	}
+
 	mountDir := info.MountDir()
+	previousActiveSymlink := filepath.Join(mountDir, "..", "previous")
+	if err := os.Remove(previousActiveSymlink); err != nil && !os.IsNotExist(err) {
+		logger.Noticef("Cannot remove %q: %v", previousActiveSymlink, err)
+	}
+	dataDir := info.DataDir()
+	previousDataSymlink := filepath.Join(dataDir, "..", "previous")
+	if err := os.Remove(previousDataSymlink); err != nil && !os.IsNotExist(err) {
+		logger.Noticef("Cannot remove %q: %v", previousDataSymlink, err)
+	}
+}
+
+func rollbackSymlinks(info *snap.Info) error {
+	if info.Type() != snap.TypeSnapd {
+		return nil
+	}
+
+	var errActiveSymlink error
+	var errDataSymlink error
 
+	mountDir := info.MountDir()
 	currentActiveSymlink := filepath.Join(mountDir, "..", "current")
+	previousActiveSymlink := filepath.Join(mountDir, "..", "previous")
 	if err := os.Remove(currentActiveSymlink); err != nil && !os.IsNotExist(err) {
 		logger.Noticef("Cannot remove %q: %v", currentActiveSymlink, err)
 	}
+	if err := os.Rename(previousActiveSymlink, currentActiveSymlink); err != nil && !os.IsNotExist(err) {
+		errActiveSymlink = fmt.Errorf("Cannot rename %q: %w", previousActiveSymlink, err)
+	}
 
 	dataDir := info.DataDir()
 	currentDataSymlink := filepath.Join(dataDir, "..", "current")
+	previousDataSymlink := filepath.Join(dataDir, "..", "previous")
 	if err := os.Remove(currentDataSymlink); err != nil && !os.IsNotExist(err) {
 		logger.Noticef("Cannot remove %q: %v", currentDataSymlink, err)
 	}
+	if err := os.Rename(previousDataSymlink, currentDataSymlink); err != nil && !os.IsNotExist(err) {
+		errDataSymlink = fmt.Errorf("Cannot rename %q: %w", previousDataSymlink, err)
+	}
+
+	if errActiveSymlink != nil {
+		return errActiveSymlink
+	}
+	return errDataSymlink
+}
+
+func updateCurrentSymlinks(info *snap.Info) (e error) {
```

Comment:
as we are changing this, should we consider usign AtomicSymlink more here?  in terms of copying current to previous we could also Readlink current to know what to put in previous instead of renaming previous? Also given that "previous" is just a backup used for a short window, maybe we should call it ".previous" ?

Reply:
Done.